src/platforms/common/stm32/dfu_f4.c

```diff
@@ -52,10 +52,11 @@ static void get_sector_num(uint32_t addr)
 {
 	if (addr < sector_addr[0])
 		return;
-	size_t i;
-	for (i = 1; sector_addr[i]; ++i) {
-		if (addr < sector_addr[i])
+	size_t i = 0;
+	while (sector_addr[i + 1]) {
```

Comment:
The for-loop transformation which we applied here is the correct thing to do - even if apparently we got either the loop condition, or more likely the post-state invariant, wrong. This is a for loop in while loop clothing.

Given what is changing here, it suggests that what was missing is really `--i;` after the for loop (line 60) so the if statement on line 62 then has the `i` value it's expecting.

Reply:
Thank you for your reply! I'edit the PR in accordance with the guidelines and test it more thoroughly (seems that `--i;` will be enough, I'll check it). Thanks for the great BMP project!

Reply:
I've tested decrement `--i;` and, as you said, it seems sufficient to find correct sector number. I've placed it right after `if (!sector_addr[i])` check so that sector number 12 would never returned from the function. As I readed from docs, sector 12 exists in stm32f4 with more than 1Mb flash memory and for that memory sectors 12-23 needed so I thought it is not necessary for now. I've attached the chunk of code where I checked the output of the function depending from input addresses so anyone can check it too. And I debugged that chunk with my hardware (blackmagic debugging blackmagic :)), DFU downloads were stable.

If I've done something wrong please correct me. it's a pleasure to be involved in the black magic project!

https://onlinegdb.com/g83JE4cmG8

![BMP-01](https://github.com/blackmagic-debug/blackmagic/assets/17275900/9b6ec058-f93f-403b-9317-0a36c2efd03e)
![BMP-02](https://github.com/blackmagic-debug/blackmagic/assets/17275900/061479c5-5704-498c-a5cd-ebe8c7f3a787)